### PR TITLE
Add Libqrcodegencpp

### PR DESCRIPTION
**Pull request by tytan652**
### Description

Some Linux distribution (e.g. Ubuntu, Fedora) tweak https://github.com/nayuki/QR-Code-generator to ship it's C and C++ implementations as libraries.

One of the methods is to add files from https://github.com/EasyCoding/qrcodegen-cmake and use CMake to build the libraries which is the method took for this PR.

The C implementation (libqrcodegen) is removed since it is not needed.

I'm aware of #174, so if it get merged first, I will update the pwsh script to add `$Targets`.

### Motivation and Context

Reduce the number of submodules in obs-studio repo and provides those as libraries in obs-deps.
Also this would allow to make obs-websocket submodule-less.

### How Has This Been Tested?

CI test build static version, and tested shared on VMs.

https://github.com/tytan652/obs-studio/pull/13 provides binaries:
 - Ubuntu 20.04 is failed too because of an API break from Libqrcodegen between 1.5.0 (20.04) and 1.7.0 (22.04) (header file name change), so one more reason to drop Ubuntu 20.04.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
I have a patch downstream in Fedora that [adapts the websocket plugin to use the system version of qrcodegencpp](https://src.fedoraproject.org/rpms/obs-studio/blob/rawhide/f/obs-studio-websocket-use-system-qrcodegencpp.patch). Once this is merged, would it be acceptable to get this upstreamed too?

**Comment by tytan652** [comment]
> I have a patch downstream in Fedora that [adapts the websocket plugin to use the system version of qrcodegencpp](https://src.fedoraproject.org/rpms/obs-studio/blob/rawhide/f/obs-studio-websocket-use-system-qrcodegencpp.patch). Once this is merged, would it be acceptable to get this upstreamed too?

Read until the end.

No, because your patch rely on installed cmake modules (made by qrcodegen-cmake) which is:
1. non-existent in Ubuntu packages
2. has a bug with a non-working namespace (`qrcodegencpp::qrcodegencpp` does not work on my testing)

I created a new finder that works for any cases that use qrcodegen.hpp. https://github.com/PatTheMav/cmake-finders/pull/2 which will be added to OBS Studio later.

And here is the branch for my changes for obs-websocket: https://github.com/tytan652/obs-websocket/tree/last_ws_dep_standing

I discovered the existence of libqrcodegencpp package through you patch and thank you for that. But it does not work in all cases, so I had create something that will work almost anywhere.

**Comment by Conan-Kudo** [comment]
@xvitaly, could you take a look at the bug @tytan652 discovered about your CMake overlay?

**Comment by tytan652** [comment]
The bug has magically disappeared because of a reboot apparently, CMake why ?.
Sorry @xvitaly for the false alarm.

**Comment by tytan652** [comment]
Sorry again, @xvitaly I can reproduce it, we are in a transition period in our CMake build-system and I modified the wrong file for Linux. I will write a issue on the qrcodegen-cmake repo rather than continuing the off-topic here.

**Comment by xvitaly** [comment]
> I will write a issue on the qrcodegen-cmake repo rather than continuing the off-topic here.

Great. I will look into it.

File: deps.windows/60-qrcodegencpp.ps1 [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    Invoke-External cmake @Options
+}
+
+function Fixup {
+    Log-Information "Fixup (${Target})"
+    Set-Location $Path
+
+    Remove-Item -ErrorAction 'SilentlyContinue' -Recurse "$($ConfigData.OutputPath)/include/qrcodegen"
```
Comment (PatTheMav) [inline deps.windows/60-qrcodegencpp.ps1:97]: Either use `Get-ChildItems` to gather the files for deletion and pipe its result to `Remove-Item` once, or generate a hash table with all the deletion jobs and pipe it to `Remove-Item`:

```
Get-ChildItem ./lib/* -Include 'qrcodegen.*' | Remove-Item -Force -Recurse

[or]

$Items | ForEach-Object {
    $Item = $_
    if ( Test-Path $Item.FullPath ) {
        Remove-Item -Recurse -Force $Item.FullPath
    }
}
```

Comment (tytan652) [reply deps.windows/60-qrcodegencpp.ps1]: Tried to do the hash table.

File: deps.windows/60-qrcodegencpp.ps1 [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    $Items = @(
+        @{ FullPath = "$($ConfigData.OutputPath)/include/qrcodegen" }
+        @{ FullPath = "$($ConfigData.OutputPath)/lib/cmake/qrcodegen" }
+        @{ FullPath = "$($ConfigData.OutputPath)/lib/pkgconfig/qrcodegen.pc" }
+        @{ FullPath = "$($ConfigData.OutputPath)/lib/qrcodegen.*" }
+        @{ FullPath = "$($ConfigData.OutputPath)/bin/qrcodegen.*" }
+    )
+
+    $Items | ForEach-Object {
+    $Item = $_
```
Comment (PatTheMav) [inline deps.windows/60-qrcodegencpp.ps1:106]: Needs indentation.

Comment (tytan652) [reply deps.windows/60-qrcodegencpp.ps1]: Sorry, I only see this review weeks later.

Fixed.

**Review by PatTheMav — APPROVED** [review]

**Comment by tytan652** [comment]
Commits:
- deps.macos: Add Libqrcodegencpp to macOS deps
- [TO SQUASH] Apply recent macOS deps changes
have been squashed since changes were approved.

**Comment by RytoEX** [comment]
I almost wish that instead of being in a single script, this were two separate scripts, like:
```
deps.macos/80-qrcodegencpp.zsh
deps.macos/81-qrcodegencmake.zsh
deps.windows/60-qrcodegencpp.ps1
deps.windows/61-qrcodegencmake.ps1
```
But I'm not sure that would necessarily be any "better" than what is already here, and it might be more convoluted to get that working, so this is fine as-is if @PatTheMav is already fine with it.

**Comment by tytan652** [comment]
> I almost wish that instead of being in a single script

`qrcodegen-cmake` is a big "patch" meant to modify the the `Qr-Code-Generator` to build as a standalone lib with CMake.
They can't be separated since we can't build `Qr-Code-Generator` without `qrcodegen-cmake`.

**Comment by PatTheMav** [comment]
That's actually an oversight on my end - the CMake files need to be added via a patch file, not via a second checkout (and using relative path patterns like `mkcd ../something` is also no bueno, instead `pushd`/`popd` pairs should be used, but when using a proper patch file to generate the required CMake files directly in the source dir this should be unnecessary anyway).

**Comment by PatTheMav** [comment]
> I will try to avoid the "relative path", but I will not increase the maintenance cost by generating a patch from a git repo meant to be copied inside another repo.

That's the canonical way we add CMake support to projects that don't have it. The other option would be a fork that has that patch applied. It seems very roundabout to me to have one repository with the sources, and another repository with the build system files, especially as you cannot combine both into the same directory.

**Comment by RytoEX** [comment]
> > I will try to avoid the "relative path", but I will not increase the maintenance cost by generating a patch from a git repo meant to be copied inside another repo.
> 
> That's the canonical way we add CMake support to projects that don't have it. The other option would be a fork that has that patch applied. It seems very roundabout to me to have one repository with the sources, and another repository with the build system files, especially as you cannot combine both into the same directory.

Unfortunately, the CMake repo itself recommends this method:
```
git clone https://github.com/nayuki/QR-Code-generator.git qrcodegen
git clone https://github.com/EasyCoding/qrcodegen-cmake.git qrcodegen-cmake
ln -s ../qrcodegen-cmake/{cmake,CMakeLists.txt} qrcodegen/
mkdir -p qrcodegen-build
cmake -S qrcodegen -B qrcodegen-build -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_EXAMPLES:BOOL=ON -DBUILD_TESTS:BOOL=ON
cmake --build qrcodegen-build
```
So as far as that project is concerned, git cloning both projects is correct.

**Comment by PatTheMav** [comment]
> > > I will try to avoid the "relative path", but I will not increase the maintenance cost by generating a patch from a git repo meant to be copied inside another repo.
> > 
> > 
> > That's the canonical way we add CMake support to projects that don't have it. The other option would be a fork that has that patch applied. It seems very roundabout to me to have one repository with the sources, and another repository with the build system files, especially as you cannot combine both into the same directory.
> 
> Unfortunately, the CMake repo itself recommends this method:
> 
> ```
> git clone https://github.com/nayuki/QR-Code-generator.git qrcodegen
> git clone https://github.com/EasyCoding/qrcodegen-cmake.git qrcodegen-cmake
> ln -s ../qrcodegen-cmake/{cmake,CMakeLists.txt} qrcodegen/
> mkdir -p qrcodegen-build
> cmake -S qrcodegen -B qrcodegen-build -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_EXAMPLES:BOOL=ON -DBUILD_TESTS:BOOL=ON
> cmake --build qrcodegen-build
> ```
> 
> So as far as that project is concerned, git cloning both projects is correct.

Yeah "one within the other" is not possible (except for downloading the tarball from GitHub and extracting it in place). I don't like it, but the implementation in the PR is fine.

**Comment by RytoEX** [comment]
> Yeah "one within the other" is not possible (except for downloading the tarball from GitHub and extracting it in place). I don't like it, but the implementation in the PR is fine.

Right. I think we can explore an alternative implementation in the future, or encourage the base project to add proper CMake support.

**Comment by xvitaly** [comment]
> The other option would be a fork that has that patch applied. It seems very roundabout to me to have one repository with the sources, and another repository with the build system files, especially as you cannot combine both into the same directory.

The QR-Code-generator upstream is too unfriendly to external contributors. They don't accept pull requests. Our PR with full CMake support: https://github.com/nayuki/QR-Code-generator/pull/140

That's why we created our own repo with CMake support.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Whitespace nit.

File: deps.windows/60-qrcodegencpp.ps1 [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+function Configure {
+    Log-Information "Configure (${Target})"
+    Set-Location "$Path/src"
+
+	$OnOff = @('OFF', 'ON')
+    $Options = @(
+        $CmakeOptions
+        "-DBUILD_SHARED_LIBS:BOOL=$($OnOff[$script:Shared.isPresent])"
+    )
```
[the remark this answers is not in the document]
Comment (tytan652) [reply deps.windows/60-qrcodegencpp.ps1]: Yes, a tab got here. Fixed.

**Review by RytoEX — APPROVED** [review]
Seems fine. Just need to determine merge order with respect to other pending PRs.

**Pull request merged**